README.md
## adhocore/twig-yall

It is a twig extension around [`malchata/yall.js`](https://github.com/malchata/yall.js)
for lazy loading `img`, `picture`, `video`, `iframe` etc.

(Also supports `source` tag and `srcset` attribute).

[![Latest Version](https://img.shields.io/github/release/adhocore/twig-yall.svg?style=flat-square)](https://github.com/adhocore/twig-yall/releases)
[![Travis Build](https://img.shields.io/travis/com/adhocore/twig-yall.svg?branch=master&style=flat-square)](https://travis-ci.com/adhocore/twig-yall?branch=master)
[![Scrutinizer CI](https://img.shields.io/scrutinizer/g/adhocore/twig-yall.svg?style=flat-square)](https://scrutinizer-ci.com/g/adhocore/twig-yall/?branch=master)
[![Codecov branch](https://img.shields.io/codecov/c/github/adhocore/twig-yall/master.svg?style=flat-square)](https://codecov.io/gh/adhocore/twig-yall)
[![StyleCI](https://styleci.io/repos/172214338/shield)](https://styleci.io/repos/172214338)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](./LICENSE)

Eager loading of images and videos not only makes webpage slow and clumsy but also consumes a lot of bandwidth.

If you use twig templating system for your view layer, this extension defers loading of resources so they are loaded only when required in viewport.

## Installation
```bash
composer require adhocore/twig-yall
```

## Usage

First setup twig to register this extension:
```php
// Use your loader of choice
$twig = new Twig\Environment(new Twig\Loader\ArrayLoader);

// Register Yall with defaults
$twig->addExtension(new Ahc\TwigYall\Yall);

// Configuring Yall instance:
$twig->addExtension(new Ahc\TwigYall\Yall(
    'polyfillJs'  => '<custom url to polyfill>',
    'yallJs'      => '<custom url to yall.js>',
    'lazyClass'   => '<default lazy class>',
    'placeholder' => '<default placeholder image url>',
));
```

Voila, then in twig templates you would either use `{% lazyload %}` block to lazyload whole block at once
OR individually lazyload each resources with `{{ lazify() }}`.

In both cases, you must call `{{ yallify() }}` somewhere at the footer.

### lazyload

With `placeholder` config set to `'default.png'`, below template
```twig
<img src="apple.jpg" />                   {# not lazyloaded #}
{% lazyload %}
<img src="ball.jpg" />                    {# lazyloaded #}
<img src="cat.jpg" class="no-lazy" />     {# not lazyloaded #}
<img src="cat.jpg" data-src="..." />      {# not lazyloaded #}
<video poster="vid.jpg">                  {# lazyloaded #}
  <source src="vid1.mp4">                 {# lazyloaded #}
  <source src="vid2.mp4">                 {# lazyloaded #}
</video>
<video class='no-lazy' src="..."></video> {# not lazyloaded #}
<picture><source src="pic.jpg"></picture> {# lazyloaded #}
{% endlazyload %}
<img src="...">                           {# not lazyloaded #}
```
will be rendered as:
```html
<img src="apple.jpg" />
<img class="lazy yall" src="default.png" data-src="ball.jpg" />
<img src="cat.jpg" class="no-lazy" />
<img src="cat.jpg" data-src="..." />
<video class="lazy yall" poster="default.png" data-poster="vid.jpg">
  <source class="lazy yall" data-src="vid1.mp4">
  <source class="lazy yall" data-src="vid2.mp4">
</video>
<video class='no-lazy' src="..."></video>
<picture><source class="lazy yall" data-src="pic.jpg"></picture>
<img src="...">
```

### lazify

#### only src
```twig
<img {{ lazify("/my/img.jpg") }} />
```
will be rendered as:
```html
<img class="lazy yall" src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEAAAAALAAAAAABAAEAAAI=" data-src="/my/img.jpg" />
```

See [stackoverflow](https://stackoverflow.com/a/15960901) for the usage of `data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEAAAAALAAAAAABAAEAAAI=`.

#### with class
```twig
<img {{ lazify("logo.png", "cls1 cls2") }} />
```
will be rendered as:
```html
<img class="cls1 cls2 lazy yall" src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEAAAAALAAAAAABAAEAAAI=" data-src="logo.png" />
```

#### custom placeholder
```twig
<img {{ lazify("logo.png", "cls1 cls2", "/img/placeholder.png") }} />
```
will be rendered as:
```html
<img class="cls1 cls2 lazy yall" src="/img/placeholder.png" data-src="logo.png" />
```

#### video poster
```twig
<video {{ lazify({poster: "video/poster.jpg"}, "video", "dummyposter.jpg") }}>
```
will be rendered as:
```html
<video class="video lazy yall" poster="dummyposter.jpg" data-poster="video/poster.jpg">
```

#### source tag
```twig
<source {{ lazify({srcset: "img2x.jpg 2x, img1x.jpg 1x"}) }}>
```
will be rendered as:
```html
<source class="lazy yall" data-srcset="img2x.jpg 2x, img1x.jpg 1x">
```

#### src+srcset
```twig
<img {{ lazify(["src.jpg", "src2x.jpg 2x", "src1x.jpg 1x"], "", "dummy.jpg") }} />
```
will be rendered as:
```html
<img class="lazy yall" src="dummy.jpg" data-src="src.jpg" data-srcset="src2x.jpg 2x, src1x.jpg 1x" />
```

### yallify

**Important:** Do not forget to put the yall loader somewhere in the footer twig template:

```twig
{{ yallify() }}
```

Which by default loads yall 3.1.7 with polyfills. You can set yall.js version, and turn off polyfill like so:
```twig
{{ yallify("3.1.6", "") }} {# load yall v3.1.6 but not polyfill #}
```

You can pass yall options in third param. For event callbacks wrap it in `<raw></raw>`:
```twig
{{ yallify(null, null, {observeChanges: true, events: {load: "<raw>function(){}</raw>"}}) }}
```
will be rendered as:
```html
<script src="https://polyfill.io/v2/polyfill.min.js?features=IntersectionObserver" async></script>
<script src="https://unpkg.com/yall-js@3.1.7/dist/yall.min.js" async></script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  window.setTimeout(function () {
    yall({
      "observeChanges": true,
      "events": {
        "load": function(){}
      },
      "lazyClass": "lazy"
    });
  }, 99);
});
</script>
```

**PS:**
The inputs sent to `lazify()` or `yallify()` are not validated by this library.

From `malchata/yall.js`:
> Use appropriate width and height attributes, styles, and lightweight placeholders for your images.

## Contributing

Please check [the guide](./CONTRIBUTING.md)

## LICENSE

> &copy; [MIT](./LICENSE) | 2019, Jitendra Adhikari


CHANGELOG.md
## [0.2.2](https://github.com/adhocore/twig-yall/releases/tag/0.2.2) (2019-12-23)

### Internal Refactors
- Add return typehint (Jitendra Adhikari) [_82644f1_](https://github.com/adhocore/twig-yall/commit/82644f1)

### Documentations
- Add more intro (Jitendra Adhikari) [_1cbb125_](https://github.com/adhocore/twig-yall/commit/1cbb125)
- Minor update (Jitendra Adhikari) [_6af6e4b_](https://github.com/adhocore/twig-yall/commit/6af6e4b)


## [0.2.1](https://github.com/adhocore/twig-yall/releases/tag/0.2.1) (2019-12-15)

### Internal Refactors
- **Parser**: Extract getSrc, merge no-lazy with data- (Jitendra Adhikari) [_9a95713_](https://github.com/adhocore/twig-yall/commit/9a95713)
- **Parser**: Extract needsSrc, merge data-src/poster check (Jitendra Adhikari) [_662c3c7_](https://github.com/adhocore/twig-yall/commit/662c3c7)


## [0.2.0](https://github.com/adhocore/twig-yall/releases/tag/0.2.0) (2019-12-15)

### Features
- Register lazyload block parser (Jitendra Adhikari) [_8414673_](https://github.com/adhocore/twig-yall/commit/8414673)
- Add lazyload block parser (Jitendra Adhikari) [_887ba05_](https://github.com/adhocore/twig-yall/commit/887ba05)

### Bug Fixes
- **Parser**: Traverse single node first (Jitendra Adhikari) [_a04dada_](https://github.com/adhocore/twig-yall/commit/a04dada)

### Internal Refactors
- **Parser**: Extract replacements, respect no-lazy, dont add src in source (Jitendra Adhikari) [_241f4a0_](https://github.com/adhocore/twig-yall/commit/241f4a0)

### Miscellaneous
- Add typehint (Jitendra Adhikari) [_c350981_](https://github.com/adhocore/twig-yall/commit/c350981)

### Documentations
- Add docblock (Jitendra Adhikari) [_68052c6_](https://github.com/adhocore/twig-yall/commit/68052c6)
- About lazyload block parser (Jitendra Adhikari) [_03c8c64_](https://github.com/adhocore/twig-yall/commit/03c8c64)


## [0.1.0](https://github.com/adhocore/twig-yall/releases/tag/0.1.0) (2019-12-14)

### Features
- Add Yall extension (Jitendra Adhikari) [_ba5ed17_](https://github.com/adhocore/twig-yall/commit/ba5ed17)

### Bug Fixes
- **Lazify**: Use lazyClass from config instead (Jitendra Adhikari) [_52106da_](https://github.com/adhocore/twig-yall/commit/52106da)
- Use list for PHP7.0 (Jitendra Adhikari) [_4fdeb46_](https://github.com/adhocore/twig-yall/commit/4fdeb46)

### Internal Refactors
- Use yall.min.js by default (Jitendra Adhikari) [_2da9b27_](https://github.com/adhocore/twig-yall/commit/2da9b27)
- **Yallify**: Make param nullable (Jitendra Adhikari) [_6b1e8e1_](https://github.com/adhocore/twig-yall/commit/6b1e8e1)

### Miscellaneous
- Add github meta files (Jitendra Adhikari) [_71ecd20_](https://github.com/adhocore/twig-yall/commit/71ecd20)
- Add project meta files (Jitendra Adhikari) [_1245c3b_](https://github.com/adhocore/twig-yall/commit/1245c3b)
- Add phpunit xml (Jitendra Adhikari) [_34f0486_](https://github.com/adhocore/twig-yall/commit/34f0486)
- Add composer.json (Jitendra Adhikari) [_8ade23e_](https://github.com/adhocore/twig-yall/commit/8ade23e)
- Add editorconfig (Jitendra Adhikari) [_b1bff51_](https://github.com/adhocore/twig-yall/commit/b1bff51)
- Update gitignore (Jitendra Adhikari) [_3ada464_](https://github.com/adhocore/twig-yall/commit/3ada464)
- Wip (Jitendra Adhikari) [_6366532_](https://github.com/adhocore/twig-yall/commit/6366532)
- Init (Jitendra Adhikari) [_6569d1c_](https://github.com/adhocore/twig-yall/commit/6569d1c)

### Documentations
- Add readme (Jitendra Adhikari) [_f01e772_](https://github.com/adhocore/twig-yall/commit/f01e772)

### Builds
- **Travis**: Retire nightly (Jitendra Adhikari) [_e0feaee_](https://github.com/adhocore/twig-yall/commit/e0feaee)
- **Ci**: Add travis yml (Jitendra Adhikari) [_1f72f6b_](https://github.com/adhocore/twig-yall/commit/1f72f6b)


## 0.0.0 2019-12-07 10:41:54 UTC

- init library


tests/YallTest.php
<?php

/*
 * This file is part of the TWIG-YALL package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

namespace Ahc\TwigYall\Test;

use Ahc\TwigYall\Yall;
use PHPUnit\Framework\TestCase;
use Twig\Environment;
use Twig\Loader\ArrayLoader;

class YallTest extends TestCase
{
    protected static $twig;

    public static function setUpBeforeClass()
    {
        static::$twig = new Environment(new ArrayLoader);

        static::$twig->addExtension(new Yall);
    }

    public function testLazify()
    {
        $this->assertSame(
            '<img class="lazy yall" src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEAAAAALAAAAAABAAEAAAI=" '
                . 'data-src="/my/img.jpg" />',
            $this->render('<img {{ lazify("/my/img.jpg") }} />'),
            'should use defaults if src only given'
        );

        $this->assertSame(
            '<img class="cls1 cls2 lazy yall" src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEAAAAALAAAAAABAAEAAAI=" '
                . 'data-src="logo.png" />',
            $this->render('<img {{ lazify("logo.png", "cls1 cls2") }} />'),
            'should append to given class'
        );

        $this->assertSame(
            '<img class="cls1 cls2 lazy yall" src="/img/placeholder.png" data-src="logo.png" />',
            $this->render('<img {{ lazify("logo.png", "cls1 cls2", "/img/placeholder.png") }} />'),
            'should use custom placeholder from 3rd param'
        );
    }

    public function testLazifyComplexSrc()
    {
        $this->assertSame(
            '<video class="video lazy yall" poster="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEAAAAALAAAAAABAAEAAAI=" '
                . 'data-poster="video/poster.jpg">',
            $this->render('<video {{ lazify({poster: "video/poster.jpg"}, "video") }}>'),
            'should support video poster'
        );

        $this->assertSame(
            '<source class="lazy yall" data-srcset="img2x.jpg 2x, img1x.jpg 1x">',
            $this->render('<source {{ lazify({srcset: "img2x.jpg 2x, img1x.jpg 1x"}) }}>'),
            'should support srcset only'
        );

        $this->assertSame(
            '<img class="lazy yall" src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEAAAAALAAAAAABAAEAAAI=" '
                . 'data-src="src.jpg" data-srcset="src2x.jpg 2x, src1x.jpg 1x" />',
            $this->render('<img {{ lazify(["src.jpg", "src2x.jpg 2x", "src1x.jpg 1x"]) }} />'),
            'should support src and srcset combo'
        );
    }

    public function testYallifyDefault()
    {
        $default = $this->render('{{ yallify() }}');

        $this->assertContains(
            'https://unpkg.com/yall-js@3.1.7/dist/yall.min.js',
            $default,
            'should contain yall.js'
        );

        $this->assertContains(
            'https://polyfill.io/v2/polyfill.min.js?features=IntersectionObserver',
            $default,
            'should contain polyfill.js'
        );

        $this->assertContains(
            'yall({"lazyClass":"lazy"})',
            $default,
            'should init yall with observe flag'
        );
    }

    public function testYallifyCustom()
    {
        $custom = $this->render('{{ yallify("3.1.6", "", {lazyClass: "lazzi", observeChanges: true}) }}');

        $this->assertContains(
            'https://unpkg.com/yall-js@3.1.6/dist/yall.min.js',
            $custom,
            'should contain yall.js 3.1.6'
        );

        $this->assertContains(
            'yall({"lazyClass":"lazzi"',
            $custom,
            'should override lazyClass option'
        );

        $this->assertContains(
            ',"observeChanges":true}',
            $custom,
            'should contain extra option'
        );

        $this->assertNotContains(
            'https://polyfill.io/v2/polyfill.min.js?features=IntersectionObserver',
            $custom,
            'shouldnt contain polyfill.js'
        );
    }

    public function testYallifyNullEmpty()
    {
        $null = $this->render('{{ yallify(null, null) }}');

        $this->assertContains(
            'https://unpkg.com/yall-js@3.1.7/dist/yall.min.js',
            $null,
            'should contain yall.js 3.1.7 if null passed'
        );

        $this->assertContains(
            'https://polyfill.io/v2/polyfill.min.js?features=IntersectionObserver',
            $null,
            'should contain polyfill v2 if null passed'
        );

        $empty = $this->render('{{ yallify("", "") }}');

        $this->assertContains(
            'https://unpkg.com/yall-js@3.1.7/dist/yall.min.js',
            $empty,
            'should contain yall.js 3.1.7 if null passed'
        );

        $this->assertNotContains(
            'https://polyfill.io/v2/polyfill.min.js?features=IntersectionObserver',
            $empty,
            'should not contain polyfill if empty passed'
        );
    }

    public function testConstruct()
    {
        $twig = new Environment(new ArrayLoader);

        $twig->addExtension(new Yall([
            'polyfillJs'  => '/js/poly.js',
            'placeholder' => '/img/default.png',
            'yallJs'      => '/js/yall.min.js',
            'lazyClass'   => 'defer',
        ]));

        $yallify = $twig->createTemplate('{{ yallify() }}')->render();

        $this->assertContains(
            '/js/poly.js',
            $yallify,
            'should contain custom polyfill'
        );

        $this->assertContains(
            '/js/yall.min.js',
            $yallify,
            'should contain custom yall js'
        );

        $this->assertContains(
            'yall({"lazyClass":"defer"})',
            $yallify,
            'should contain custom lazyClass'
        );

        $this->assertSame(
            'class="defer yall" src="/img/default.png" data-src="/cat.jpeg"',
            $twig->createTemplate('{{ lazify("/cat.jpeg") }}')->render(),
            'should contain custom placeholder and lazyClass'
        );
    }

    public function testYallifyEvents()
    {
        $this->assertContains(
            '{"events":{"load":fnname,"unload":function(){}}',
            $this->render('{{ yallify(0, 0, {events: {load: "<raw>fnname</raw>", unload: "<raw>function(){}</raw>"}}) }}'),
            'should contain events as raw js (not json string)'
        );
    }

    protected function render(string $template): string
    {
        return static::$twig->createTemplate($template)->render();
    }
}


tests/bootstrap.php
<?php

/*
 * This file is part of the TWIG-YALL package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

require_once __DIR__ . '/../vendor/autoload.php';


tests/ParserTest.php
<?php

/*
 * This file is part of the TWIG-YALL package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

namespace Ahc\TwigYall\Test;

use Ahc\TwigYall\Yall;
use PHPUnit\Framework\TestCase;
use Twig\Environment;
use Twig\Loader\ArrayLoader;

class ParserTest extends TestCase
{
    protected static $twig;

    public static function setUpBeforeClass()
    {
        static::$twig = new Environment(new ArrayLoader);

        static::$twig->addExtension(new Yall([
            'placeholder' => 'img/default.png',
            'lazyClass'   => 'defer',
        ]));
    }

    public function testLazyloadSimple()
    {
        $simple = $this->render('
            {% lazyload %}
                <img src="apple.jpg" />
                <img src="ball.jpg" />
            {% endlazyload %}
        ');

        $this->assertContains(
            '<img class="defer yall" src="img/default.png" data-src="apple.jpg" />',
            $simple,
            'should defer image'
        );

        $this->assertContains(
            '<img class="defer yall" src="img/default.png" data-src="ball.jpg" />',
            $simple,
            'should defer image'
        );
    }

    public function testLazyLoadComplex()
    {
        $complex = $this->render('
            {% lazyload %}
              {% if true %}
                <img src="apple.jpg" />
              {% else %}
                <img src="ball.jpg" />
              {% endif %}
            {% endlazyload %}
        ');

        $this->assertContains(
            '<img class="defer yall" src="img/default.png" data-src="apple.jpg" />',
            $complex,
            'should defer image'
        );

        $this->assertNotContains(
            '<img class="defer yall" src="img/default.png" data-src="ball.jpg" />',
            $complex,
            'should defer image'
        );

        $this->assertNotContains(
            '<img src="ball.jpg" />',
            $complex,
            'should not defer image when not met'
        );
    }

    public function testLazyLoadMixed()
    {
        $mixed = $this->render('
            <img src="above.jpg" />
            {% lazyload %}
              <img src="inner.jpg" />
              {% if true %}
                <img src="apple.jpg" />
              {% else %}
                <img src="ball.jpg" />
              {% endif %}
              <img {{ lazify("lazify.jpg") }} />
            {% endlazyload %}
            <img src="below.jpg" />
        ');

        $this->assertContains(
            '<img src="above.jpg" />',
            $mixed,
            'should not defer image above lazyload block'
        );

        $this->assertContains(
            '<img src="below.jpg" />',
            $mixed,
            'should not defer image below lazyload block'
        );

        $this->assertContains(
            '<img class="defer yall" src="img/default.png" data-src="inner.jpg" />',
            $mixed,
            'should defer image inside block'
        );

        $this->assertContains(
            '<img class="defer yall" src="img/default.png" data-src="apple.jpg" />',
            $mixed,
            'should defer image when condition met'
        );

        $this->assertContains(
            '<img class="defer yall" src="img/default.png" data-src="lazify.jpg" />',
            $mixed,
            'should defer image already using lazify'
        );
    }

    public function testLazyLoadNoDefer()
    {
        $irregular = "
            <picture><source><img/></picture>
            <img class='no-defer' src='no-lazy.jpg' />
            <img data-src='already-lazy.jpg' />
            <source data-srcset='img1.jpg 1x, img2.jpg 2x'>
            <video data-poster='poster'>
        ";

        $this->assertContains(
            trim($irregular),
            $this->render("{% lazyload %}$irregular{% endlazyload %}"),
            'shouldnot defer empty or already deferred or no-defer enforced'
        );
    }

    public function testLazyloadNonImg()
    {
        $nonImg = $this->render("
            {% lazyload %}
            <video src='vid.mp4'></video>
            <video poster='vid.jpg'>
              <source src='vid1.mp4'>
              <source src='vid2.mp4'>
            </video>
            {% endlazyload %}
        ");

        $this->assertContains(
            '<video class="defer yall" src="img/default.png" data-src=\'vid.mp4\'></video>',
            $nonImg,
            'should defer video'
        );

        $this->assertContains(
            '<video class="defer yall" poster="img/default.png" data-poster=\'vid.jpg\'>',
            $nonImg,
            'should defer video[poster]'
        );

        $this->assertContains(
            "<source class=\"defer yall\" data-src='vid1.mp4'>",
            $nonImg,
            'should defer video.source'
        );

        $this->assertContains(
            "<source class=\"defer yall\" data-src='vid2.mp4'>",
            $nonImg,
            'should defer video.source'
        );
    }

    protected function render(string $template): string
    {
        return static::$twig->createTemplate($template)->render();
    }
}


src/Parser.php
<?php

/*
 * This file is part of the TWIG-YALL package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

namespace Ahc\TwigYall;

use Twig\Node\Node;
use Twig\Node\TextNode;
use Twig\Token;
use Twig\TokenParser\AbstractTokenParser;

class Parser extends AbstractTokenParser
{
    /** @var string yall lazyClass */
    protected $lazyClass;

    /** @var string URI for placeholder image */
    protected $placeholder;

    /**
     * Constructor.
     *
     * @param string $lazyClass
     * @param string $placeholder
     */
    public function __construct(string $lazyClass, string $placeholder)
    {
        $this->lazyClass   = $lazyClass;
        $this->placeholder = $placeholder;
    }

    /**
     * Parse `{% lazyload %}...{% endlazyload %}` block.
     *
     * @param Token $token
     *
     * @return Node
     */
    public function parse(Token $token): Node
    {
        $stream = $this->parser->getStream();

        $stream->expect(Token::BLOCK_END_TYPE);
        $body = $this->parser->subparse([$this, 'isLazyloadEnd'], true);
        $stream->expect(Token::BLOCK_END_TYPE);

        return $this->traverse($body);
    }

    protected function traverse(Node $node): Node
    {
        if ($node instanceof TextNode) {
            $node->setAttribute('data', $this->doLazyload($node->getAttribute('data')));
        }

        foreach ($node as $sub) {
            $this->traverse($sub);
        }

        return $node;
    }

    protected function doLazyload(string $html): string
    {
        return \preg_replace_callback('/<(img|source|video)([^>]+)>/m', function ($match) {
            list($all, $tag, $props) = $match;

            if (empty(\trim($props, ' /'))) {
                return $all;
            }

            // Already there or no-flagged
            if (\preg_match('/(data-(src|poster)|no-' . $this->lazyClass . ')/i', $props)) {
                return $all;
            }

            $src = $this->getSrc($tag, $props);
            if (\stripos($props, ' class') === false) {
                $tag .= " class=\"{$this->lazyClass} yall\"";
            }
            if (\stripos($props, ' poster') !== false) {
                $tag .= " poster=\"{$this->placeholder}\"";
            }

            return "<$tag$src" . $this->doReplacements($props) . '>';
        }, $html);
    }

    protected function getSrc(string $tag, string $props): string
    {
        // For source no need src
        if ($tag === 'source') {
            return '';
        }

        // For video no need src if not already there!
        if ($tag === 'video' && \stripos($props, 'src') === false) {
            return '';
        }

        return ' src="' . $this->placeholder . '"';
    }

    protected function doReplacements(string $props): string
    {
        $replacements = [
            ' src='      => ' data-src=',
            ' src ='     => ' data-src=',
            ' srcset='   => ' data-srcset=',
            ' srcset ='  => ' data-srcset=',
            ' class="'   => " class=\"{$this->lazyClass} yall ",
            ' class ="'  => " class=\"{$this->lazyClass} yall ",
            ' class = "' => " class=\"{$this->lazyClass} yall ",
            " class='"   => " class='{$this->lazyClass} yall ",
            " class ='"  => " class='{$this->lazyClass} yall ",
            " class = '" => " class='{$this->lazyClass} yall ",
            ' poster='   => ' data-poster=',
            ' poster ='  => ' data-poster=',
        ];

        return \strtr($props, $replacements);
    }

    /**
     * @internal
     */
    public function isLazyloadEnd(Token $token): bool
    {
        return $token->test('endlazyload');
    }

    /**
     * Gets the tag name used in block.
     *
     * @return string
     */
    public function getTag(): string
    {
        return 'lazyload';
    }
}


src/Yall.php
<?php

/*
 * This file is part of the TWIG-YALL package.
 *
 * (c) Jitendra Adhikari <jiten.adhikary@gmail.com>
 *     <https://github.com/adhocore>
 *
 * Licensed under MIT license.
 */

namespace Ahc\TwigYall;

use Twig\Extension\AbstractExtension;
use Twig\Markup;
use Twig\TwigFunction;

class Yall extends AbstractExtension
{
    /** @var array Configuration */
    protected $config = [];

    /**
     * Constructor.
     *
     * @param array $config Optinal configuration
     */
    public function __construct(array $config = [])
    {
        $this->config = $config + [
            'polyfillJs'  => 'https://polyfill.io/%s/polyfill.min.js?features=IntersectionObserver',
            // @see: https://stackoverflow.com/a/15960901
            'placeholder' => 'data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEAAAAALAAAAAABAAEAAAI=',
            'yallJs'      => 'https://unpkg.com/yall-js@%s/dist/yall.min.js',
            'lazyClass'   => 'lazy',
        ];
    }

    /**
     * Get twig functions defined by this extension.
     *
     * @return array
     */
    public function getFunctions(): array
    {
        return [
            new TwigFunction('lazify', [$this, 'lazify']),
            new TwigFunction('yallify', [$this, 'yallify']),
        ];
    }

    /**
     * Get token parsers defined by this extension.
     *
     * @return array
     */
    public function getTokenParsers(): array
    {
        return [new Parser($this->config['lazyClass'], $this->config['placeholder'])];
    }

    /**
     * Loads yall and polyfill scripts then triggers lazy loading.
     *
     * @param string|null $yall     Yall version
     * @param string|null $polyfill Polyfill version ('' = off)
     * @param array       $options  Options for `yall({})` callback
     *
     * @return Markup
     */
    public function yallify(string $yall = null, string $polyfill = null, array $options = []): Markup
    {
        $yallJs   = \sprintf($this->config['yallJs'], $yall ?: '3.1.7');
        $options += ['lazyClass' => $this->config['lazyClass']];
        $jsonFlag = \JSON_UNESCAPED_SLASHES | \JSON_FORCE_OBJECT;

        $jsonOpts = \json_encode($options, $jsonFlag);
        $jsonOpts = \str_replace(['"<raw>', '</raw>"'], ['', ''], $jsonOpts);

        $markup = [
            $polyfill ?? 'v2'
                ? \sprintf('<script src="%s" async></script>', \sprintf($this->config['polyfillJs'], $polyfill ?? 'v2'))
                : '',
            \sprintf('<script src="%s" async></script>', $yallJs),
            '<script type="text/javascript">',
            'document.addEventListener("DOMContentLoaded", function() {',
            \sprintf('  window.setTimeout(function () { yall(%s); }, 99);', $jsonOpts),
            '});',
            '</script>',
        ];

        return new Markup(\implode("\n", $markup), 'UTF-8');
    }

    /**
     * Lazify resources.
     *
     * @param string|string[] $src     The sources to lazy load
     * @param string          $classes The optional element classes
     * @param string          $dummy   The optional placeholder image
     *
     * @return Markup
     */
    public function lazify($src, string $classes = '', string $dummy = ''): Markup
    {
        $attr = 'src';
        if (\is_array($src)) {
            list($attr, $src) = $this->normalizeSrc($src);
        }

        $classes = \trim("$classes {$this->config['lazyClass']} yall");
        if ('srcset' !== $attr) {
            $dummy = \sprintf(' %s="%s"', $attr, $dummy ?: $this->config['placeholder']);
        }

        $markup = \sprintf('class="%s"%s data-%s="%s"', $classes, $dummy, $attr, $src);

        return new Markup($markup, 'UTF-8');
    }

    protected function normalizeSrc(array $src): array
    {
        if ($src['poster'] ?? false) {
            return ['poster', $src['poster']];
        }

        if ($src['srcset'] ?? false) {
            return ['srcset', $src['srcset']];
        }

        $srcset = $src;
        $src    = \array_shift($srcset);

        return ['src', $src . '" data-srcset="' . \implode(', ', $srcset)];
    }
}
